NIP-40 Expiration Timestamp (#87)
nostr-protocol/nips: 40.md

NIP-40
======

Expiration Timestamp
-----------------------------------

`draft` `optional` `author:0xtlt`

> The `expiration` tag enables users to specify a timestamp at which the message should be considered expired and deleted by relays. This helps ensure that messages are not retained by the network after they are no longer relevant, which helps maintain the privacy of the messages. This is especially useful in cases where a replacement event is not available or feasible.

Reviewer: I would not justify this nip with privacy. We all know that events once broadcast will get stored for eternity.

```suggestion
The `expiration` tag enables users to specify a timestamp at which the message should be considered expired and deleted by relays. This helps reduce the burden of the network around events with relevance limited in time. This is especially useful in cases where a replacement event is not available or feasible.
```

Reviewer 2: I concur with Giszmo and ok300. I cannot imagine how nostr could create any reasonable expectation of privacy outside of anonymity. Once a post is sent relays can do anything with it, and clients can cache it forever (as mine currently does).

But i ♥ this nip for a different reason: relays will fill up if they don't eventually drop irrelevant events. I've seen relay code that just drops everything after a certain time period (a relay that I cannot use for a blog), and this NIP offers a useful hint to the relay.  I would expire my 'tweets' after a month, and expire my blog posts after 3 years, for example.

#### Spec

```
tag: expiration
options:

>  - [timestamp]: required

Reviewer: ```suggestion
 - [UNIX timestamp in seconds]: required
```

```

#### Example

```json
{
    "pubkey": "<pub-key>",
    "created_at": 1000000000,
    "kind": 1,
    "tags": [
      ["expiration", 1600000000]
    ],
    "content": "This message will expire at the specified timestamp and be deleted by relays.\n",
    "id": "<event-id>"
}
```

Note: The timestamp should be in the same format as the created_at timestamp and should be interpreted as the time at which the message should be deleted by relays.

Client Behavior
---------------

> Clients SHOULD use the `supported_nips` field to learn if a relay supports this NIP. Clients SHOULD NOT send expiration events to relays that do not support this NIP.

Reviewer: Relays oblivious to this nip will not care about such desires and just mirror these events. A nip can only impose stuff on software supporting the nip, not others. I wouldn't even imply there were any effort to keep these events away from relays that don't support the nip.

Author: Was already "implemented" for nip16 : https://github.com/nostr-protocol/nips/blob/master/16.md

But I understand what you have in mind.

It's not a requirement but a recommendation here

Suggested Use Cases
-------------------

* Temporary announcements - This tag can be used to make temporary announcements that are automatically deleted from the network after a certain amount of time. For example, an event organizer could use this tag to post announcements about an upcoming event, and the announcements will be automatically deleted once the event is over.
* Limited-time offers - This tag can be used by businesses to make limited-time offers that expire after a certain amount of time. For example, a business could use this tag to make a special offer that is only available for a limited time, and the offer will be automatically deleted from the network once the time has passed.

> * Ephemeral messages - This tag can be used by individuals to send ephemeral messages that are automatically deleted from the network after a certain amount of time. For example, someone could use this tag to send a message that is only visible to the recipient for a short time, and the message will be automatically deleted once the time has passed. This can be useful for sending sensitive information that should not be retained on the network indefinitely.

Reviewer: Don't broadcast sensitive information! Ever!

Author: So I am also responding here with the example of a signal-like application with custom relays

Reviewer: In a group chat, in a DM, don't ever send information you require to not be available after a certain expiry date. It simply can't be enforced and Signal pretending it could is a disservice. With the very varied client landscape in nostr there is zero chance of all client agreeing to delete expired events. If information being linkable to your account after time X is a problem, don't sign and send it before time X neither.

NIP-40
======

Expiration Timestamp
-----------------------------------

`draft` `optional` `author:0xtlt`

The `expiration` tag enables users to specify a timestamp at UTC-0 at which the message should be considered expired and deleted by relays (or just ignored by clients). This can be useful for making temporary announcements, making limited-time offers, or sending ephemeral messages but should not be seen as a security solution for the content of your messages.
So be careful when sharing confidential information.

#### Spec

```
tag: expiration
options:
 - [UNIX timestamp in seconds]: required
```

#### Example

```json
{
    "pubkey": "<pub-key>",
    "created_at": 1000000000,
    "kind": 1,
    "tags": [
      ["expiration", 1600000000]

>     ],

Author: Good idea the ttl, however, the advantage of the timestamp here is that the relay can easily locate and delete / not send or treat differently a message with an outdated timestamp.
It is also possible to do this by adding the timestamp of the creation date + the ttl but I am for a Nostr philosophy where relays are very simple programs and clients are explicit in their requests.

It would avoid a relay to make a big loop + calculation in the messages to delete them for example.

    "content": "This message will expire at the specified timestamp and be deleted by relays.\n",
    "id": "<event-id>"
}
```

Note: The timestamp should be in the same format as the created_at timestamp and should be interpreted as the time at which the message should be deleted by relays.

Client Behavior
---------------

> Clients SHOULD use the `supported_nips` field to learn if a relay supports this NIP. Clients SHOULD NOT send expiration events to relays that do not support this NIP.

Author: Even if the relay can store the message indefinitely and/or transfer it to relays that store it infinitely, I prefer to leave this line in the PIN.

This will not be the first nip to include a rule like this

Clients MUST ignore events that have expired.

> 

Author: As the expiration time must be at UTC-0, the local times is not a problem here

> 

Reviewer 3: @x1ddos said
> As the expiration time must be at UTC-0, the local times is not a problem here

There is no such thing as UTC-0 in a decentralized network.

Services that rely on an accurate time, like 2FA or TOTP, can run into issues because of time drift:

* https://blog.gaborszathmari.me/tricking-google-authenticator-totp-with-ntp/
* https://www.reddit.com/r/synology/comments/f9lu53/2fa_and_ntp_woes/
* https://forum.matomo.org/t/2fa-authentication-suddenly-fails/45829

Basing message deletion on that might be asking for trouble :)

Suggested Use Cases
-------------------

* Temporary announcements - This tag can be used to make temporary announcements that are automatically deleted from the network after a certain amount of time. For example, an event organizer could use this tag to post announcements about an upcoming event, and the announcements will be automatically deleted once the event is over.

> * Limited-time offers - This tag can be used by businesses to make limited-time offers that expire after a certain amount of time. For example, a business could use this tag to make a special offer that is only available for a limited time, and the offer will be automatically deleted from the network once the time has passed.

Reviewer 3: ```suggestion
* Limited-time offers - This tag can be used by businesses to make limited-time offers that expire after a certain amount of time. For example, a business could use this tag to make a special offer that is only available for a limited time.
```

* Ephemeral messages - This tag can be used by individuals to send ephemeral messages that are automatically deleted from the network after a certain amount of time. For example, someone could use this tag to send a message that is only visible to the recipient for a short time, and the message will be automatically deleted by relays or/and ignored by clients once the time has passed. This can be useful for sensitive information that should not be retained on the network indefinitely.

#### One last time, be careful!
The events could be downloaded by third parties as they are publicly accessible all the time on the relays.

> So don't consider ephemeral messages as a security for your conversations or other uses.

Reviewer 3: ```suggestion
So don't consider expiring messages as a security feature for your conversations or other uses.
```

#### Spec

```
tag: expiration
values:
 - [UNIX timestamp in seconds]: required
```

#### Example

```json
{
    "pubkey": "<pub-key>",
    "created_at": 1000000000,
    "kind": 1,
    "tags": [
      ["expiration", "1600000000"]
    ],
    "content": "This message will expire at the specified timestamp and be deleted by relays.\n",
    "id": "<event-id>"
}
```

Note: The timestamp should be in the same format as the created_at timestamp and should be interpreted as the time at which the message should be deleted by relays.

Client Behavior
---------------

Clients SHOULD use the `supported_nips` field to learn if a relay supports this NIP. Clients SHOULD NOT send expiration events to relays that do not support this NIP.

Clients SHOULD ignore events that have expired.

Relay Behavior
--------------

Relays MAY NOT delete an expired message immediately on expiration and MAY persist them indefinitely.  
Relays SHOULD NOT send expired events to clients, even if they are stored.  
Relays SHOULD drop any events that are published to them if they are expired.  

> Relays SHOULD persist ephemeral events until their expiration timestamp if an ephemeral event has one, and relays MAY limit the amount of time an ephemeral event is persisted.

Reviewer: Ephemeral events currently do not get persisted. They get relaid to interested parties and dropped. With this change, they turn into non-ephemeral events. I'm against expiration for ephemeral events. If anything I would make this explicit:

```suggestion
An expiration timestamp does not affect storage of ephemeral events.
```
